synthesizer/process/src/lib.rs

```diff
@@ -173,48 +177,65 @@ impl<N: Network> Process<N> {
             .chain(std::iter::once((*stack.program_id(), stack))) // add the root stack.
             .unique_by(|(id, _)|*id) // don't add duplicates.
             .filter(|(program_id, _)| {
-                program_id != &ProgramID::<N>::from_str("credits.aleo").unwrap() // don't add the credits.aleo stack.
+                *program_id != credits_id // don't add the credits.aleo stack.
                 && !stacks.contains(program_id) // don't add stacks present in the cache.
             })
             .collect::<Vec<_>>();
         // Determine the required capacity.
         let remaining_capacity = stacks.cap().get().saturating_sub(stacks.len());
-        let mut required_capacity = programs_to_add.len().saturating_sub(current_capacity);
+        let mut required_capacity = programs_to_add.len().saturating_sub(remaining_capacity);
         if required_capacity != 0 {
             debug!("Evicting {required_capacity} stacks from the cache.");
         }
         // If the new stacks require more capacity, attempt to remove the least recently used stacks.
         while required_capacity != 0 {
-            // To avoid dangling stacks, we only remove stacks that do not have more than 1 reference (e.g. as external stack).
-            // For this not to loop endlessly, we assume to always quickly find stacks with just one reference, so:
-            // 1. queried stacks must be short-lived, which is essential anyway to avoid memory leaks.
-            // 2. queried stacks must expire before a querying thread queries the next stack.
+            // To avoid dangling stacks, we only remove stacks which are not imported by any other stack.
             let root_program_ids = stacks
                 .iter()
                 .rev()
-                .filter_map(|(program_id, stack)| (std::sync::Arc::<_>::strong_count(stack) == 1).then_some(program_id))
+                .filter_map(|(program_id, (_, num_parents))| (*num_parents == 0).then_some(program_id))
                 .take(required_capacity)
                 .cloned()
                 .collect::<Vec<_>>();
```

Comment:
this should not need to be collected anymore

Reply:
The iterator is holding (immutable) references to a part of `stacks`. As long as part of the iterator lives, I cannot call any `stacks.pop(&program_id)`.

Reply:
ah, that's fair; I thought we only needed to collect it to have the count 👍